### ISSUE TITLE
Research: supporting using hive in multiple isolates

### ISSUE THREAD

**Issue by themisir** [hivedb/hive#922]
## Why?

People are using Hive in background isolates to work with persistent data, and because hive doesn't have proper isolate support, and dart doesn't have proper thread support, it's currently impossible to make sure different hive instances can sync up with each other when mutating database. So proper isolate support was requested multiple times on [hive/hive issues](https://github.com/hivedb/hive/issues?q=is%3Aissue+is%3Aopen+isolate).

## When?

I don't know, it totally depends on our findings on this research thread.

## Limitations

- We can't send lots of data through isolates, even tho isolates are improved a lot, it's still costly a lot for workloads like database
- Dart doesn't have proper (fully compatible) [Thread](https://en.wikipedia.org/wiki/Thread_%28computing%29) API
- We have to keep existing API and database structure as stable as possible

## Other references

- Some comments on dart-lang threading support issue: [1](https://github.com/dart-lang/language/issues/333#issuecomment-1060825399), [2](https://github.com/dart-lang/language/issues/333#issuecomment-1061619521)
- [IsolateNameServer](https://api.flutter.dev/flutter/dart-ui/IsolateNameServer-class.html)

**Comment by TheCarpetMerchant** [comment]
Hey ! @mtc-jed here.

Here's a possible implementation that didn't pan out, but it taught me a lot about Isolates. It uses `dart:ui`'s `IsolateNameServer` but would need a real mutex for security in deciding which Isolate Hive lives in.
To note : I wrote this in while Dart 2.15 was in beta I believe, with some flags activated for constructor tear-offs and Isolates Groups. These things are now on by default, so this actually might work as is.

I had trouble with the Isolate not responding, possibly due to the way Isolates are handled (Isolate Groups could have fixed this issue) and the permissions they have (for file access ; you can't access the app's files if the Isolate wasn't created by the app, again probably a thing now handled by Isolate Groups).

I wanted to use this to be able to access Hive from the Isolate spawned by Android's AlarmManager. It would be worth asking that team if that Isolate is part of the same Group as the main app's Isolate in all circumstances, as I believe this to be the crux of the problem with that use case.

Anyway, here goes : https://github.com/TheCarpetMerchant/HiveIsolateInterface/blob/master/HiveIsolateInterface.dart

Had to work around Hive a bit to make it work. Hope the comments make it understandable what's going on.

**Comment by themisir** [comment]
I have a few concerns about your implementations. For example:

1. IsolateA spawned from notification or alarm manager and created master hive instance
2. User opens notification, OS opens app and main / UI isolate sees that there's a alive hive port so uses it
3. OS decides to stop notification since user tapped on it and notification cleared, isolate gets killed, but hive still thinks there's master hive instance on another isolate. 

So to solve that hive might need to periodically do health checks to decide which isolate has to process data which might cause race condition again (if both isolates tries to be master isolate at same time).

---

Btw I was actually thinking about utilizing ports as like event bus. So instead of sending data to master hive instance, when a hive instance makes some change it notifies other hive instances to reload box (or mark that box as dirty so next time user needs to read something it has to be reloaded). But still this has race condition issue, which could be solved using mutexes.

**Comment by TheCarpetMerchant** [comment]
What you're describing is exactly why we'd need to talk to the team handling the AlarmManager plugin, so we know how these things are behaving.
I thought about calling the `isAlive` function before every call but that would be tremendously costly.

The other approach you're describing is very interesting, and might be easily implemented at least for LazyBoxes. I believe LazyBoxes only read the keys and corresponding offset to the data in the file ? So you'd only need to re-read that.

The real problem is how to implement a correct mutex for this in Dart. Could it be possible to handle this is Java instead ?

**Comment by themisir** [comment]
> The real problem is how to implement a correct mutex for this in Dart. Could it be possible to handle this in Java instead ?

Yeah that's another option. Actually I thought about writing whole communication part (comms between hive instances) on low level language (to provide cross-platform support). We'll also not gonna have limitations on dart like shared memory space or handling race conditions. But I don't know if it does worth it. It'll probably create another set of issues.

**Comment by TheCarpetMerchant** [comment]
I think a first goal should be to provide multi-Isolate support for existing uses (ie Flutter apps), if it happens to be feasible without too much trouble, as this would unlock many use cases for Hive, such as the AlarmManager stuff. But if it happens to be too much trouble handling this in Dart itself, side-stepping Dart's limitations through `dart:ffi` is the only viable option. Maybe a solid Rust library could provide safe shared memory management ?

**Comment by themisir** [comment]
> But if it happens to be too much trouble handling this in Dart itself, side-stepping Dart's limitations through dart:ffi is the only viable option. Maybe a solid Rust library could provide safe shared memory management ?

Yeah, I think that's why @leisim used rust instead of dart for writing core of [isar](https://isar.dev).

**Comment by TheCarpetMerchant** [comment]
A simple solution could be to load the entire Hive file as-is in Rust-managed shared memory. Then instead of Hive reading/writing things to the disk, it's writing to that shared memory through Rust, with Rust handling mutexes.
I'm not sure how easily you could "update" the Hive instances from there though.
Maybe it's better to have a specific Box type for this shared memory system, where objects are read and deserialized from that shared memory every time we need it ?

**Comment by GColumbu** [comment]
@themisir any updates on this? I have exactly this problem...I am updating a hive box on a background task and want to sync up with the main isolate when I resume the application.
